src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,273 @@
+                lock (_dictLock)
+                {
+                    if (!lineToTokensDict.TryGetValue(currentLineNumber, out lineTokens))
+                    {
+                        if (tokens.Count != 0)
+                        {
+                            // We already have some sort of cached results, but we've now reached
+                            // a gap in the cache which we have no tokens for.
+                            // For ease of computation, we only want to return a continuous range.
+                            // We'll keep our current set of cached results and exit the loop early.
+                            break;
+                        }
+
+                        // We don't have cached tokens for the line. We'll check other lines in the
+                        // requested range to see if we might have cached tokens for those.
+                        continue;
+                    }
+                }
+
+                cachedRangeStart ??= new Position { Line = currentLineNumber, Character = 0 };
+                cachedRangeLength++;
+
+                // If a line's List is empty, it means the line has no tokens. However, we still want
+                // to keep track of empty lines in order to later compute line offsets.
+                if (lineTokens.Count == 0)
+                {
+                    currentEmptyLineCount++;
+                    continue;
+                }
+
+                // We've found a non-empty cached line. Go through and process it.
+                ProcessCachedLine(tokens, currentEmptyLineCount, currentLineNumber, lineTokens);
+
+                currentEmptyLineCount = 0;
+            }
+
+            return tokens;
+        }
+
+        private static List<int> ProcessCachedLine(
+            List<int> tokens,
+            int currentEmptyLineCount,
+            int currentLineNumber,
+            IReadOnlyList<int> lineTokens)
+        {
+            // If we're at the start of the cached range (i.e. we've found no prior tokens in
+            // the range up til now), the line offset should be relative to the start of the
+            // document.
+            if (tokens.Count == 0)
+            {
+                tokens.Add(currentLineNumber);
+            }
+            // If we've already found prior cached tokens in the requested range, the line
+            // offset should be relative to the end of the last cached non-empty line.
+            // Example 1:
+            //     token1
+            //     token2
+            // `token2` should have a line offset of 1.
+            // Example 2:
+            //     token1
+            //     [empty line]
+            //     [empty line]
+            //     token2
+            // `token2` should have a line offset of 3 since we need to take empty lines
+            // into account.
+            else
+            {
+                tokens.Add(1 + currentEmptyLineCount);
```

Comment:
i genuinely don't see how 1+currentEmptyLineCount corresponds to the above comment.

Reply:
ah, wait.  i get it now.  ok that comment is helpful.